src/search3_indexer.erl

```diff
@@ -0,0 +1,221 @@
+            proc => Proc
+        },
+        try
+            true = proc_prompt(Proc, [<<"add_fun">>, Index#index.def]),
+            update_int(Db, NewState)
+        after
+            ret_os_process(Proc)
+        end
+    catch error:database_does_not_exist ->
+        #{db_prefix := DbPrefix} = Db,
+        couch_log:error("Db - ~p database does not exist", [DbPrefix]),
+        erlang:error(database_does_not_exist)
+    end.
+
+update_int(#{} = Db, State) ->
+    State3 = fabric2_fdb:transactional(Db, fun(TxDb) ->
+        State1 = maps:put(tx_db, TxDb, State),
+        {ok, State2} = fold_changes(State1),
+        #{
+            count := Count,
+            limit := Limit,
+            doc_acc := DocAcc,
+            last_seq := LastSeq,
+            index := Index,
+            proc := Proc
+        } = State2,
+
+        % 1) purge_seq is empty for now, subsequent releases will support this
+        % 2) We are indexing inside a transaction, which has a timeout of 5s.
+        % This can potentially be a problem for large documents. We should
+        % revisit this design later.  
+        index_docs(Index, Proc, LastSeq, <<>>, DocAcc),
+        case Count < Limit of
+            true ->
+                report_progress(State2, finished),
+                finished;
+            false ->
+                report_progress(State2, update),
+                State2#{
+                    tx_db := undefined,
+                    count := 0,
+                    doc_acc := [],
+                    search_seq => LastSeq,
+                    % make sure this reset here is correct
+                    last_seq := 0
+                }
+        end
+    end),
+    case State3 of
+        finished ->
+            % should we ret_os_process(Proc) here? or in the after clause
+            % in update/2?
+            ok;
+        _ ->
+            update_int(Db, State3)
+    end.
+
+fold_changes(State) ->
+    #{
+        search_seq := SearchSeq,
+        limit := Limit,
+        tx_db := TxDb
+    } = State,
+    fabric2_db:fold_changes(TxDb, SearchSeq,
+        fun load_changes/2, State, [{limit, Limit}]).
+
+% grabs the document from changes feed, and stores it into the document
+% accumulator for indexing later
+load_changes(Change, Acc) ->
+    #{
+        doc_acc := DocAcc,
+        count := Count,
+        tx_db := TxDb
+    } = Acc,
+    #{
+        id := Id,
+        sequence := LastSeq,
+        deleted := Deleted
+    } = Change,
+    Acc1 = case Id of
+        <<"_design/", _/binary>> ->
+            maps:merge(Acc, #{
+                count => Count + 1,
+                last_seq => LastSeq
+                });
+        _ ->
+            Doc = if Deleted -> []; true ->
+                case fabric2_db:open_doc(TxDb, Id) of
+                    {ok, Doc0} -> Doc0;
+                    {not_found, _} -> []
+                end
+            end,
+            Change1 = maps:put(doc, Doc, Change),
+            maps:merge(Acc, #{
```

Comment:
Could use syntactic update (instead of maps:put and also instead of merge?). Something like:

```
Acc#{
   doc_acc := DocAcc ++ [Change#{doc := Doc}],
   count := Count +1,
   last_seq := LastSeq
}
```